File: clients/native/android/app/src/main/java/ai/wuu/nativeapp/AccountPush.kt
package ai.wuu.nativeapp

import org.json.JSONObject

/** Consent belongs to one login, including its bearer-token lifetime. */
object PushConsent {
    fun key(session: AccountSession) = sha256(listOf(session.server, session.username, session.pub, session.token).joinToString("\n").toByteArray()).b64()
    fun encode(session: AccountSession, enabled: Boolean) = json("login" to key(session), "enabled" to enabled)
    fun matches(session: AccountSession, consent: JSONObject) = consent.optString("login") == key(session)
}

suspend fun AccountAPI.pushStatus(token: String) = request("/push", token)
suspend fun AccountAPI.setPush(token: String, platform: String, deviceToken: String?) {
    if (deviceToken != null) {
        require(deviceToken.isNotEmpty() && platform in setOf("ios", "android")) { "Invalid push registration" }
        request("/push", token, json("platform" to platform, "token" to deviceToken))
    } else request("/push", token, method = "DELETE")
}

File: clients/native/ios/Sources/WuuCore/AccountPush.swift
import Foundation
import CryptoKit

/// Consent belongs to one login, including its bearer-token lifetime.
public struct PushConsent: Codable, Sendable {
    private let login: String
    public let enabled: Bool
    private static func key(_ session: AccountSession) -> String {
        Data(SHA256.hash(data: Data([session.server, session.username, session.pub, session.token].joined(separator: "\n").utf8))).base64URL
    }
    public init(session: AccountSession, enabled: Bool) { login = Self.key(session); self.enabled = enabled }
    public func matches(_ session: AccountSession) -> Bool { login == Self.key(session) }
}

public struct AccountPushStatus: Decodable, Sendable {
    public let enabled: Bool
    public let platform: String
}

extension AccountAPI {
    public func pushStatus(token: String) async throws -> AccountPushStatus { try await request("/push", token: token) }
    public func setPush(token: String, platform: String, deviceToken: String?) async throws {
        if let deviceToken {
            guard !deviceToken.isEmpty, ["ios", "android"].contains(platform) else { throw NativeError.invalid("Invalid push registration") }
            let _: JSONValue = try await request("/push", token: token, body: ["platform": platform, "token": deviceToken])
        } else {
            let _: JSONValue = try await request("/push", token: token, method: "DELETE")
        }
    }
}
